components/micro-gateway-core/src/main/ballerina/src/gateway/jwt_generator/claim_retriever.bal

```diff
@@ -0,0 +1,136 @@
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/jwt;
+import ballerina/runtime;
+
+boolean claimRetrieveEnabled = true;
+
+# To retrieve claims via the user specific claim retrieve implementation.
+# 
+# + userInfo - Authentication Context of the user, which is provided as input to the claim retriever Implementation
+# + return - ClaimListDTO if there are any claims added from the user specific implementation
+function retrieveClaims (UserAuthContextDTO? userInfo) returns @tainted RetrievedUserClaimsListDTO ? {
+    //if claim retrieve variable is disabled, there is no need to run through the method.
+    if (!claimRetrieveEnabled) {
+        return;
+    }
+    if (userInfo is UserAuthContextDTO) {
+        printDebug (CLAIM_RETRIEVER, "User Auth Context information provided to the claim retrieval implementation : " +
+                    userInfo.toString());
+        RetrievedUserClaimsListDTO? | error claimListDTO = trap retrieveClaimsFromImpl(userInfo);
+        if (claimListDTO is RetrievedUserClaimsListDTO ) {
+            printDebug (CLAIM_RETRIEVER, "Claims List received from the claim retrieval implementation : " +
+                        claimListDTO.toString());
+            return claimListDTO;
+        } else if (claimListDTO is ()) {
+            printDebug(CLAIM_RETRIEVER , "No user claims are received from the claim retrieval implementation");
+        } else {
+            printError(CLAIM_RETRIEVER , "Error while retrieving user claims from the claim retrieval implementation",
+                claimListDTO);
+        }
+    }
+}
+
+# To do the class loading operation for the user specific claim retriever implementation.
+public function loadClaimRetrieverImpl() {
+    
+    //todo: bring a configuration if required
+    if (!isConfigAvailable(JWT_GENERATOR__CLAIM_RETRIEVAL_INSTANCE_ID, JWT_GENERATOR_CLAIM_RETRIEVAL_IMPLEMENTATION)) {
+        printDebug(CLAIM_RETRIEVER, "Claim Retrieval related class loading is disabled as the implementation is not provided." +  
+                    "Hence claim retrieval is disabled");
+        claimRetrieveEnabled = false;            
+        return;
+    }
+
+    string claimRetrieverImplClassName = getConfigValue(JWT_GENERATOR__CLAIM_RETRIEVAL_INSTANCE_ID,
+                                                        JWT_GENERATOR_CLAIM_RETRIEVAL_IMPLEMENTATION,
+                                                        DEFAULT_JWT_GENERATOR_CLAIM_RETRIEVAL_IMPLEMENTATION);
+    map<any> claimRetrieverConfig = getConfigMapValue(JWT_GENERATOR_CLAIM_RETRIEVAL_CONFIGURATION);
+
+    if (claimRetrieverConfig.length() == 0) {
```

Comment:
No need to check if map is empty, to provide user the capability to add additional properties to the existing apim credentials